cartesian_controllers/src/twist_controller.cpp

```diff
@@ -0,0 +1,149 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "cartesian_controllers/twist_controller.hpp"
+
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "controller_interface/helpers.hpp"
+#include "hardware_interface/loaned_command_interface.hpp"
+
+namespace cartesian_controllers
+{
+using hardware_interface::LoanedCommandInterface;
+
+TwistController::TwistController()
+: controller_interface::ControllerInterface(),
+  rt_command_ptr_(nullptr),
+  twist_command_subscriber_(nullptr)
+{
+}
+
+controller_interface::InterfaceConfiguration TwistController::command_interface_configuration()
+  const
+{
+  controller_interface::InterfaceConfiguration command_interfaces_config;
+  command_interfaces_config.type = controller_interface::interface_configuration_type::INDIVIDUAL;
+
+  for (const auto & interface : interface_names_)
+  {
+    command_interfaces_config.names.push_back(joint_name_ + "/" + interface);
+  }
+
+  return command_interfaces_config;
+}
+
+controller_interface::InterfaceConfiguration TwistController::state_interface_configuration() const
+{
+  return controller_interface::InterfaceConfiguration{
+    controller_interface::interface_configuration_type::NONE};
+}
+
+CallbackReturn TwistController::on_init()
+{
+  try
+  {
+    auto_declare<std::vector<std::string>>("interface_names", std::vector<std::string>());
+
+    auto_declare<std::string>("joint", "");
+  }
+  catch (const std::exception & e)
+  {
+    fprintf(stderr, "Exception thrown during init stage with message: %s \n", e.what());
+    return CallbackReturn::ERROR;
+  }
+
+  return CallbackReturn::SUCCESS;
+}
+
+CallbackReturn TwistController::on_configure(const rclcpp_lifecycle::State & /*previous_state*/)
+{
+  joint_name_ = get_node()->get_parameter("joint").as_string();
+
+  if (joint_name_.empty())
+  {
+    RCLCPP_ERROR(get_node()->get_logger(), "'joint' parameter was empty");
+    return CallbackReturn::ERROR;
+  }
+
+  // Specialized, child controllers set interfaces before calling configure function.
+  if (interface_names_.empty())
+  {
+    interface_names_ = get_node()->get_parameter("interface_names").as_string_array();
+  }
+
+  if (interface_names_.empty())
+  {
+    RCLCPP_ERROR(get_node()->get_logger(), "'interface_names' parameter was empty");
+    return CallbackReturn::ERROR;
+  }
+
+  twist_command_subscriber_ = get_node()->create_subscription<CmdType>(
+    "~/commands", rclcpp::SystemDefaultsQoS(),
```

Comment:
We use `reference` now. See [this file](https://github.com/ros-controls/roadmap/blob/master/design_drafts/design_of_controllers_and_their_status.md)

```suggestion
    "~/reference", rclcpp::SystemDefaultsQoS(),
```